File: tools/release/publish_swiftpm_source_tag.py
#!/usr/bin/env python3
"""Publish or verify the semver source tag SwiftPM needs for the Apple SDK."""

from __future__ import annotations

import argparse
import os
import re
import subprocess
import sys
import tempfile
from pathlib import Path
from typing import NoReturn

import product_metadata


ROOT = Path(__file__).resolve().parents[2]
SEMVER_RE = re.compile(
    r"^(0|[1-9][0-9]*)[.](0|[1-9][0-9]*)[.](0|[1-9][0-9]*)(?:[-+][0-9A-Za-z.-]+)?$"
)


def fail(message: str) -> NoReturn:
    print(f"publish_swiftpm_source_tag.py: {message}", file=sys.stderr)
    raise SystemExit(1)


def git_output(args: list[str]) -> str:
    return subprocess.check_output(["git", *args], cwd=ROOT, text=True).strip()


def git_run(args: list[str], *, env: dict[str, str] | None = None) -> None:
    subprocess.run(["git", *args], cwd=ROOT, env=env, check=True)


def commit_for_ref(ref: str) -> str:
    return git_output(["rev-parse", f"{ref}^{{commit}}"])


def tag_ref(tag: str) -> str:
    return f"refs/tags/{tag}"


def tag_commit(tag: str) -> str | None:
    result = subprocess.run(
        ["git", "rev-parse", "--verify", "--quiet", f"{tag_ref(tag)}^{{commit}}"],
        cwd=ROOT,
        check=False,
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.DEVNULL,
    )
    if result.returncode == 0:
        return result.stdout.strip()
    return None


def swiftpm_tag() -> str:
    version = product_metadata.read_current_version("oliphaunt-swift")
    if SEMVER_RE.fullmatch(version) is None:
        fail(f"SwiftPM requires a semantic version tag; oliphaunt-swift version is {version!r}")
    return version


def commit_parents(commit: str) -> list[str]:
    parts = git_output(["rev-list", "--parents", "-n", "1", commit]).split()
    return parts[1:]


def file_at_ref(ref: str, path: str) -> str | None:
    result = subprocess.run(
        ["git", "show", f"{ref}:{path}"],
        cwd=ROOT,
        check=False,
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.DEVNULL,
    )
    return result.stdout if result.returncode == 0 else None


def tree_for_commit(commit: str) -> str:
    return git_output(["rev-parse", f"{commit}^{{tree}}"])


def synthetic_commit_matches(commit: str, parent: str, expected_tree: str) -> bool:
    return commit_parents(commit) == [parent] and tree_for_commit(commit) == expected_tree


def iter_tree_files(root: Path) -> list[Path]:
    files: list[Path] = []
    for path in sorted(root.rglob("*")):
        if path.is_file():
            files.append(path)
        elif not path.is_dir():
            fail(f"SwiftPM generated release tree contains unsupported file type: {path}")
    return files


def add_blob_to_index(env: dict[str, str], path: str, data: str | bytes) -> None:
    binary = isinstance(data, bytes)
    blob_output = subprocess.run(
        ["git", "hash-object", "-w", "--stdin"],
        cwd=ROOT,
        env=env,
        check=True,
        text=not binary,
        input=data,
        stdout=subprocess.PIPE,
    ).stdout
    blob = blob_output.decode("utf-8").strip() if binary else blob_output.strip()
    git_run(["update-index", "--add", "--cacheinfo", f"100644,{blob},{path}"], env=env)


def create_swiftpm_release_tree(
    target_commit: str,
    manifest: str,
    include_trees: list[Path],
) -> str:
    base_tree = git_output(["rev-parse", f"{target_commit}^{{tree}}"])
    with tempfile.TemporaryDirectory(prefix="oliphaunt-swiftpm-index.") as tmp:
        env = {**os.environ, "GIT_INDEX_FILE": str(Path(tmp) / "index")}
        git_run(["read-tree", base_tree], env=env)
        add_blob_to_index(env, "Package.swift", manifest)
        for include_tree in include_trees:
            root = include_tree.resolve()
            if not root.is_dir():
                fail(f"SwiftPM generated release tree does not exist: {include_tree}")
            for file in iter_tree_files(root):
                relative = file.relative_to(root).as_posix()
                if relative == "Package.swift" or relative.startswith(".git/") or "/.git/" in relative:
                    fail(f"SwiftPM generated release tree contains forbidden path: {relative}")
                add_blob_to_index(env, relative, file.read_bytes())
        return subprocess.run(
            ["git", "write-tree"],
            cwd=ROOT,
            env=env,
            check=True,
            text=True,
            stdout=subprocess.PIPE,
        ).stdout.strip()


def create_swiftpm_manifest_commit(target_commit: str, tree: str, version: str) -> str:
    return subprocess.run(
        [
            "git",
            "commit-tree",
            tree,
            "-p",
            target_commit,
            "-m",
            f"Release Oliphaunt Swift {version} SwiftPM manifest",
        ],
        cwd=ROOT,
        check=True,
        text=True,
        stdout=subprocess.PIPE,
    ).stdout.strip()


def ensure_tag(target: str, *, manifest_path: str | None, include_trees: list[str], push: bool) -> str:
    tag = swiftpm_tag()
    version = product_metadata.read_current_version("oliphaunt-swift")
    target_commit = commit_for_ref(target)
    manifest = None
    tag_target = target_commit
    expected_tree = tree_for_commit(target_commit)

    if manifest_path is not None:
        manifest = (ROOT / manifest_path).read_text(encoding="utf-8")
        if "binaryTarget(" not in manifest or "liboliphaunt-native-v" not in manifest:
            fail("SwiftPM release manifest must contain a checksum-pinned liboliphaunt binaryTarget")
        expected_tree = create_swiftpm_release_tree(
            target_commit,
            manifest,
            [(ROOT / include_tree) for include_tree in include_trees],
        )
        tag_target = create_swiftpm_manifest_commit(target_commit, expected_tree, version)

    existing = tag_commit(tag)
    if existing is not None:
        if manifest is not None and synthetic_commit_matches(existing, target_commit, expected_tree):
            print(f"SwiftPM version tag {tag} already points at a release manifest commit for {target_commit}")
            tag_target = existing
        elif existing != tag_target:
            fail(
                f"SwiftPM version tag {tag} already points at {existing}, "
                f"not expected SwiftPM release commit {tag_target}"
            )
        else:
            print(f"SwiftPM version tag {tag} already points at {tag_target}")
    else:
        git_run(["tag", tag, tag_target])
        print(f"created SwiftPM version tag {tag} at {tag_target}")

    if push:
        git_run(["push", "origin", tag_ref(tag)])
        print(f"pushed SwiftPM version tag {tag} to origin")
    return tag


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--target",
        default=os.environ.get("GITHUB_SHA", "HEAD"),
        help="commitish that the SwiftPM version tag must derive from",
    )
    parser.add_argument(
        "--manifest",
        help=(
            "generated public SwiftPM Package.swift to place in a release-only "
            "tag commit; when omitted, the semver tag points directly at --target"
        ),
    )
    parser.add_argument(
        "--include-tree",
        action="append",
        default=[],
        help=(
            "generated repository-relative file tree to include in the release-only "
            "SwiftPM tag commit; may be passed multiple times"
        ),
    )
    parser.add_argument(
        "--push",
        action="store_true",
        help="push the tag to origin after creating or verifying it locally",
    )
    return parser.parse_args(argv)


def main(argv: list[str]) -> int:
    args = parse_args(argv)
    ensure_tag(args.target, manifest_path=args.manifest, include_trees=args.include_tree, push=args.push)
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: tools/release/publish_swiftpm_source_tag.mjs
#!/usr/bin/env bun
import { spawnSync } from "node:child_process";
import {
  mkdtempSync,
  readdirSync,
  readFileSync,
  rmSync,
  statSync,
} from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";

import { currentVersion } from "./product-version.mjs";

const ROOT = path.resolve(import.meta.dir, "../..");
const SEMVER_RE = /^(0|[1-9][0-9]*)[.](0|[1-9][0-9]*)[.](0|[1-9][0-9]*)(?:[-+][0-9A-Za-z.-]+)?$/u;
const FIRST_OLIPHAUNT_SWIFTPM_VERSION = [0, 6, 0];
const RELEASE_BOT_NAME = "oliphaunt-release-bot";
const RELEASE_BOT_EMAIL = "oliphaunt-release-bot@users.noreply.github.com";
const decoder = new TextDecoder();

function fail(message) {
  console.error(`publish_swiftpm_source_tag.mjs: ${message}`);
  process.exit(1);
}

function usage(status = 1) {
  const message =
    "usage: tools/release/publish_swiftpm_source_tag.mjs [--target COMMITISH] [--manifest PACKAGE_SWIFT] [--include-tree TREE]... [--push]";
  if (status === 0) {
    console.log(message);
    process.exit(0);
  }
  fail(message);
}

function valueArg(argv, index, name) {
  const value = argv[index + 1];
  if (value === undefined || value.startsWith("--")) {
    fail(`${name} requires a value`);
  }
  return value;
}

function parseArgs(argv) {
  const args = {
    target: process.env.GITHUB_SHA || "HEAD",
    manifest: undefined,
    includeTrees: [],
    push: false,
  };
  for (let index = 0; index < argv.length; ) {
    const arg = argv[index];
    if (arg === "--target") {
      args.target = valueArg(argv, index, arg);
      index += 2;
    } else if (arg === "--manifest") {
      args.manifest = valueArg(argv, index, arg);
      index += 2;
    } else if (arg === "--include-tree") {
      args.includeTrees.push(valueArg(argv, index, arg));
      index += 2;
    } else if (arg === "--push") {
      args.push = true;
      index += 1;
    } else if (arg === "--help" || arg === "-h") {
      usage(0);
    } else {
      usage();
    }
  }
  if (!args.target) {
    fail("--target must not be empty");
  }
  return args;
}

function git(args, { root = ROOT, env = process.env, check = true, input = undefined } = {}) {
  const result = spawnSync("git", args, {
    cwd: root,
    env,
    input,
    encoding: input instanceof Buffer ? "buffer" : "utf8",
    stdout: "pipe",
    stderr: "pipe",
  });
  if (check && result.status !== 0) {
    const stderr = Buffer.isBuffer(result.stderr)
      ? decoder.decode(result.stderr).trim()
      : String(result.stderr).trim();
    fail(`git ${args.join(" ")} failed${stderr ? `: ${stderr}` : ""}`);
  }
  const stdout = Buffer.isBuffer(result.stdout)
    ? decoder.decode(result.stdout)
    : String(result.stdout);
  return {
    status: result.status ?? 0,
    stdout: stdout.trim(),
  };
}

function commitForRef(ref, root) {
  return git(["rev-parse", `${ref}^{commit}`], { root }).stdout;
}

function tagRef(tag) {
  return `refs/tags/${tag}`;
}

function tagCommit(tag, root) {
  const result = git(["rev-parse", "--verify", "--quiet", `${tagRef(tag)}^{commit}`], {
    root,
    check: false,
  });
  return result.status === 0 ? result.stdout : null;
}

function stableVersionCore(version) {
  const match = SEMVER_RE.exec(version);
  return match === null ? null : match.slice(1, 4).map(Number);
}

function compareVersionCore(left, right) {
  for (let index = 0; index < left.length; index += 1) {
    if (left[index] !== right[index]) {
      return left[index] - right[index];
    }
  }
  return 0;
}

async function swiftpmTag(versionOverride) {
  const version = versionOverride ?? await currentVersion("oliphaunt-swift");
  if (!SEMVER_RE.test(version)) {
    fail(`SwiftPM requires a semantic version tag; oliphaunt-swift version is ${JSON.stringify(version)}`);
  }
  if (compareVersionCore(stableVersionCore(version), FIRST_OLIPHAUNT_SWIFTPM_VERSION) < 0) {
    fail(
      `SwiftPM version tag ${version} collides with the legacy unscoped tag range; ` +
        "the first Oliphaunt SwiftPM version is 0.6.0",
    );
  }
  return version;
}

function commitParents(commit, root) {
  const parts = git(["rev-list", "--parents", "-n", "1", commit], { root }).stdout.split(/\s+/u).filter(Boolean);
  return parts.slice(1);
}

function treeForCommit(commit, root) {
  return git(["rev-parse", `${commit}^{tree}`], { root }).stdout;
}

function syntheticCommitMatches(commit, parent, expectedTree, root) {
  const parents = commitParents(commit, root);
  return parents.length === 1 && parents[0] === parent && treeForCommit(commit, root) === expectedTree;
}

function iterTreeFiles(root) {
  const files = [];
  function visit(directory) {
    for (const entry of readdirSync(directory, { withFileTypes: true }).sort((a, b) => a.name.localeCompare(b.name))) {
      const file = path.join(directory, entry.name);
      if (entry.isDirectory()) {
        visit(file);
      } else if (entry.isFile()) {
        files.push(file);
      } else {
        fail(`SwiftPM generated release tree contains unsupported file type: ${file}`);
      }
    }
  }
  visit(root);
  return files.sort();
}

function addBlobToIndex(root, env, indexPath, data) {
  const result = git(["hash-object", "-w", "--stdin"], { root, env, input: data });
  git(["update-index", "--add", "--cacheinfo", `100644,${result.stdout},${indexPath}`], { root, env });
}

export function createSwiftpmReleaseTree(targetCommit, manifest, includeTrees, { root = ROOT } = {}) {
  const baseTree = treeForCommit(targetCommit, root);
  const tempRoot = mkdtempSync(path.join(tmpdir(), "oliphaunt-swiftpm-index."));
  try {
    const env = { ...process.env, GIT_INDEX_FILE: path.join(tempRoot, "index") };
    git(["read-tree", baseTree], { root, env });
    addBlobToIndex(root, env, "Package.swift", manifest);
    for (const includeTree of includeTrees) {
      const includeRoot = path.resolve(root, includeTree);
      if (!statSync(includeRoot, { throwIfNoEntry: false })?.isDirectory()) {
        fail(`SwiftPM generated release tree does not exist: ${includeTree}`);
      }
      for (const file of iterTreeFiles(includeRoot)) {
        const relative = path.relative(includeRoot, file).split(path.sep).join("/");
        if (relative === "Package.swift" || relative.startsWith(".git/") || relative.includes("/.git/")) {
          fail(`SwiftPM generated release tree contains forbidden path: ${relative}`);
        }
        addBlobToIndex(root, env, relative, readFileSync(file));
      }
    }
    return git(["write-tree"], { root, env }).stdout;
  } finally {
    rmSync(tempRoot, { recursive: true, force: true });
  }
}

function commitTimestamp(commit, root) {
  const timestamp = git(["show", "-s", "--format=%ct", commit], { root }).stdout;
  if (!/^[0-9]+$/u.test(timestamp)) {
    fail(`could not derive a deterministic timestamp from release commit ${commit}`);
  }
  return `${timestamp} +0000`;
}

export function createSwiftpmManifestCommit(
  targetCommit,
  tree,
  version,
  { root = ROOT, ambientEnv = process.env } = {},
) {
  const date = commitTimestamp(targetCommit, root);
  const env = {
    ...ambientEnv,
    GIT_AUTHOR_NAME: RELEASE_BOT_NAME,
    GIT_AUTHOR_EMAIL: RELEASE_BOT_EMAIL,
    GIT_AUTHOR_DATE: date,
    GIT_COMMITTER_NAME: RELEASE_BOT_NAME,
    GIT_COMMITTER_EMAIL: RELEASE_BOT_EMAIL,
    GIT_COMMITTER_DATE: date,
  };
  return git([
    "commit-tree",
    tree,
    "-p",
    targetCommit,
    "-m",
    `Release Oliphaunt Swift ${version} SwiftPM manifest`,
  ], { root, env }).stdout;
}

export async function ensureTag(
  { target, manifest, includeTrees, push },
  { root = ROOT, version: versionOverride } = {},
) {
  const version = await swiftpmTag(versionOverride);
  const tag = version;
  const targetCommit = commitForRef(target, root);
  let tagTarget = targetCommit;
  let expectedTree = treeForCommit(targetCommit, root);
  let manifestText = null;

  if (manifest !== undefined) {
    manifestText = readFileSync(path.resolve(root, manifest), "utf8");
    if (!manifestText.includes("binaryTarget(") || !manifestText.includes("liboliphaunt-native-v")) {
      fail("SwiftPM release manifest must contain a checksum-pinned liboliphaunt binaryTarget");
    }
    expectedTree = createSwiftpmReleaseTree(targetCommit, manifestText, includeTrees, { root });
    tagTarget = createSwiftpmManifestCommit(targetCommit, expectedTree, version, { root });
  }

  const existing = tagCommit(tag, root);
  if (existing !== null) {
    if (manifestText !== null && syntheticCommitMatches(existing, targetCommit, expectedTree, root)) {
      console.log(`SwiftPM version tag ${tag} already points at a release manifest commit for ${targetCommit}`);
      tagTarget = existing;
    } else if (existing !== tagTarget) {
      fail(`SwiftPM version tag ${tag} already points at ${existing}, not expected SwiftPM release commit ${tagTarget}`);
    } else {
      console.log(`SwiftPM version tag ${tag} already points at ${tagTarget}`);
    }
  } else {
    git(["tag", tag, tagTarget], { root });
    console.log(`created SwiftPM version tag ${tag} at ${tagTarget}`);
  }

  if (push) {
    git(["push", "origin", tagRef(tag)], { root });
    console.log(`pushed SwiftPM version tag ${tag} to origin`);
  }
  return tag;
}

if (import.meta.main) {
  await ensureTag(parseArgs(Bun.argv.slice(2)));
}
